### PR TITLE
stage2: variable shadowing detection

### PR DESCRIPTION
**Pull request by g-w1**
This allows detection of variable shadowing in stage2, fixing a TODO in `varDecl` function. I added a `compileError` test to test/stage2/test.zig, but it didn't seem to run.

Edit:
The test actually did run and pass, but it said that it ran the wrong number of tests?
```
[jacob@archlinux ~co/zig/zig]$ zig build test-stage2
Test [1/17] test "self-hosted"... tests [1/28] referencing decls which appear later in the file [1/1] updAll 17 tests passed.
[jacob@archlinux ~co/zig/zig]$
```
Why does it say 1/17 and 1/28? and then only 17 tests pass? I think this is a bug in the testing code.

### CODE REVIEW

**Comment by EleanorNB** [comment]
Nice.

File: test/stage2/test.zig [hunk 1]
```diff
@@ -1112,6 +1112,24 @@ pub fn addCases(ctx: *TestContext) !void {
...
+            \\    unreachable;
+            \\}
+        , &[_][]const u8{":3:9: error: redeclaration of variable 'i'"});
+        case.addError(
+            \\var testing: i64 = 10;
+            \\export fn _start() noreturn {
+            \\    var testing: i64 = 20;
+            \\    unreachable;
+            \\}
+        , &[_][]const u8{":3:9: error: redeclaration of variable 'testing'"});
```
Comment (Vexu) [inline test/stage2/test.zig:1130]: Stage1 would give `redefinition of 'testing'` for this but I think that should be made the error for both cases.

**Comment by LemonBoy** [comment]
This implementation has the same two-step scope lookup as stage1, including the flaw noticed in #7148.
I'd expect the stage2 compiler to implement the correct shadowing logic, or are we playing the "It doesn't have to be perfect" card?

**Comment by g-w1** [comment]
I just used the logic from the identifier resolution code, so if we want to implement the more "advanced" lookup logic we should probably abstract it into a "lookup" function and then use it in both places.

https://github.com/ziglang/zig/blob/master/src/astgen.zig#L1901-L1925

**Comment by LemonBoy** [comment]
It's not a "basic" vs "more advanced" thing, it's a "according to the specification" vs "stage1's behaviour" case.
Keep in mind that a lot of code will be broken by this change, don't forget to buckle up.

**Comment by Vexu** [comment]
The problem with stage1 is that declarations only look for redefinitions in the same container, not with local variables.

**Comment by andrewrk** [comment]
The logic looks OK to me here. Let me clarify the language specification (cc @SpexGuy):

Declarations are allowed to collide, so that it is possible to declare, for example, both of these:

  * `a.b.c`
  * `a.c`

```zig
pub const a = struct {
    pub const b = struct {
        pub const c = struct {};
    };
    pub const c = struct {};
};
```

This is allowed, however, ambiguous references to `c` would be a compile error. So, inside `b`, one must use `a.c` or `b.c` in order to refer to `c`. Referring to `c` directly would be a compile error.

Parameters and local variables are never allowed to shadow anything, including each other, or any declarations, no exceptions.

**Comment by LemonBoy** [comment]
So, if I understand correctly your example, #7140 is legal but the use of `print` inside `print` should trigger an error?

**Comment by andrewrk** [comment]
Yes

**Comment by andrewrk** [comment]
#678

**Pull request merged**